system-addon/lib/TopSitesFeed.jsm

```diff
@@ -54,30 +62,45 @@ this.TopSitesFeed = class TopSitesFeed {
     let screenshot = await Screenshots.getScreenshotForURL(url);
     const action = {type: at.SCREENSHOT_UPDATED, data: {url, screenshot}};
     this.store.dispatch(ac.BroadcastToContent(action));
+    return screenshot;
+  }
+
+  /**
+   * Migrate cached link data by copying over screenshots.
+   */
+  _linkMigrator(oldLink, newLink) {
+    for (const property of ["fetchingScreenshot", "screenshot"]) {
+      const oldValue = oldLink[property];
+      if (oldValue) {
+        newLink[property] = oldValue;
+      }
+    }
   }
   async getLinksWithDefaults(action) {
     // Get at least SHOWMORE amount so toggling between 1 and 2 rows has sites
     const numItems = Math.max(this.store.getState().Prefs.values.topSitesCount,
       TOP_SITES_SHOWMORE_LENGTH);
-    let frecent = await NewTabUtils.activityStreamLinks.getTopSites({numItems});
-    const notBlockedDefaultSites = DEFAULT_TOP_SITES.filter(site => !NewTabUtils.blockedLinks.isBlocked({url: site.url}));
-    const defaultUrls = notBlockedDefaultSites.map(site => site.url);
-    let pinned = this._getPinnedWithData(frecent);
-    pinned = pinned.map(site => site && Object.assign({}, site, {
-      isDefault: defaultUrls.indexOf(site.url) !== -1,
-      hostname: shortURL(site)
-    }));
+    const frecent = (await this.frecentCache.request({
+      numItems,
+      topsiteFrecency: FRECENCY_THRESHOLD
+    })).map(link => Object.assign(link, {hostname: shortURL(link)}));
```

Comment:
Again, I generally don't like the idea of assuming no side effects will occur if you mutate stuff that comes from other functions. It relies on the particular implementation of the cache to return copies, and if that were changed, we'd have to remember to check/update this code

Reply:
How would you update the cache? We could expose a method on `LinksCache` to update an entry, but either we need to search for the entry on each update or keep around a mapping for lookup?

Reply:
I suppose it would be similar to migrating from old link to new link.

Reply:
Added a `{}` as the first agument to the `assign` here and for pinned too.